clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -223,59 +303,130 @@ public String getScheme() {
+	if (i == -1) {
+	    loc.setRef(s);
+	} else {
+	    loc.setRef(s.substring(0, i));
+	    loc.setPath(s.substring(i+1));
+	}
+	return loc;
+    }
+
+    private static String trimLeadingSlash(String s) {
+	if (s.startsWith(SEPARATOR)) {
+	    return s.substring(1);
+	}
+	return s;
     }
 
     private static class ObjectLocation {
-        private String repository;
-        private String ref;
-        private String path;
-        
-        public String getRepository() {
-            return repository;
-        }
-
-        public void setRepository(String repository) {
-            this.repository = repository;
-        }
-
-        public String getRef() {
-            return ref;
-        }
-
-        public void setRef(String ref) {
-            this.ref = ref;
-        }
-
-        public String getPath() {
-            return path;
-        }
-
-        public void setPath(String path) {
-            this.path = path;
-        }
+	private String repository;
+	private String ref;
+	private String path;
+
+	public String getRepository() {
+	    return repository;
+	}
+
+	public void setRepository(String repository) {
+	    this.repository = repository;
+	}
+
+	public String getRef() {
+	    return ref;
+	}
+
+	public void setRef(String ref) {
+	    this.ref = ref;
+	}
+
+	public String getPath() {
+	    return path;
+	}
+
+	public void setPath(String path) {
+	    this.path = path;
+	}
+    }
+
+    @Override
+    public boolean exists(Path f) throws IOException {
+	return false;
+    }
+
+    /**
+     * Wraps a FSDataOutputStream to link file on staging when done writing
+     */
+    static private class LinkOnCloseOutputStream extends java.io.OutputStream {
+	private AmazonS3 s3Client;
+	private StagingApi staging;
+	private StagingLocation loc;
+	private ObjectLocation objLoc;
+	private URI physicalUri;
+	private java.io.OutputStream out;
+
+	LinkOnCloseOutputStream(AmazonS3 s3Client, StagingApi staging, StagingLocation loc, ObjectLocation objLoc, URI physicalUri, java.io.OutputStream out) {
+	    this.s3Client = s3Client;
+	    this.staging = staging;
+	    this.loc = loc;
+	    this.objLoc = objLoc;
+	    this.physicalUri = physicalUri;
+	    this.out = out;
+	}
+
+	@Override
+	public void flush() throws IOException {
+	    out.flush();
+	}
+
+	@Override
+	public void write(byte[] b) throws IOException {
+	    out.write(b);
+	}
+
+	@Override
+	public void write(byte[] b, int off, int len) throws IOException {
+	    out.write(b, off, len);
+	}
+
+	@Override
+	public void write(int b) throws IOException {
+	    out.write(b);
+	}
+
+	@Override
+	public void close() throws IOException {
+	    out.close();
+	    // Now the object is on the underlying store, find its parameters (sadly lost by
+	    // the underlying Hadoop FileSystem) so we can link it on lakeFS.
+	    String bucket = physicalUri.getHost();
+	    String key = trimLeadingSlash(physicalUri.getPath());
+	    ObjectMetadata res = s3Client.getObjectMetadata(bucket, key);
```

Comment:
Do you mind renaming res to a more descriptive name?

Reply:
Sure.  `res` is often a shortcut for "result" of an API call, but agreed that this is not particularly descriptive.